rlog/2023-09-27-gossipimprovements.mdx

```diff
@@ -0,0 +1,214 @@
+The aspect of dissimilar capabilities and resource-constrained nodes is less explored.
+
+It is discussed in GOAL1 that overlay mesh results in better performance if $D_{avg} = \ln(N) + C$. 
+Enforcing all the nodes to have approximately $\ln(N) + C$ peers makes resource-rich nodes under-utilized, while resource-constrained nodes are overloaded. 
+At the same time, connecting high-bandwidth nodes through a low-bandwidth node undermines the network's performance.
+Ideally, every node should contribute proportionally to the available resources.
+A better solution involves a two-phased operation:
+
+
+1. Every node computes its available bandwidth and selects a node degree $D_i$ proportional to its available bandwidth [3].
+
+Different bandwidth estimation approaches are suggested in literature [4,5]. 
+
+We suggest using a simple bandwidth estimation approach similar to the ones used by dynamic adaptive video streaming over HTTP (DASH) [6]. 
+
+To further conform to the suggested mesh-degree average $D_{avg}$, every node tries achieving this number within its neighborhood, resulting in an overall similar mesh-degree average. 
+
+2. From the available local view, every node tries connecting peers with the lowest latency until $D$ connections are made.
+We suggest referring to the peering solution discussed in GOAL5 to avoid partitioning.
+
+The current GossipSub design considers homogeneous peers, and every node tries maintaining $D_{Low} \leq D \leq D_{High}$ connections. 
+
+
+## GOAL3: Bandwidth Maximization
+Redundant message transmissions are essential for handling adversaries/node failure. However, these transmissions result in traffic bursts, cramming many overlay links. 
+This not only adds to the network-wide message dissemination latency but a significant share of the network's bandwidth is wasted on (usually) unnecessary transmissions. 
+It is essential to explore solutions that can minimize the impact of redundant transmissions while assuring resilience against node failures. 
+
+Many efforts have been made to minimize the impact of redundant transmissions. 
+These solutions include multicast delivery trees, metadata sharing to enable pull-based operation, in-network information caching, etc. [7,8]. 
+GossipSub employs a hybrid of eager push (message dissemination through the overlay) and lazy push (a pull-based operation by the nodes requiring information through IWANT messages). 
+
+A better alternative is to use the redundant transmissions as a critical advantage for the GossipSub protocol. 
+Suppose that we have three equal-length messages $x1, x2, x3$. Assuming an XOR coding function,
+we know two trivial properties: $x1 \oplus x2 \oplus x2 = x1$ and $\vert x1 \vert = \vert x1 \oplus x2 \oplus x2 \vert$. 
+
+This implies that instead of sending messages individually, we can encode and transmit composite message(s) to the network. 

```

Comment:
This also implies that the messages are coming from the same node, which is only one specific use case

Reply:
I think a protocol can be devised (maybe by adding peer-IDs and message hash/ID-s) to indicate combined messages.

Reply:
ba67fe604acb1211a2617f461adc08b8097f1e47